### PR TITLE
DB detail editing permission enforcement

### PR DESCRIPTION
**Pull request by noahmoss**
Updates the `Database` model so that a write check will pass if a user has DB details permissions for the database.

Updates relevant endpoints to do write checks, rather than superuser checks.

Adds new tests and modifies some existing tests.

### CODE REVIEW

File: src/metabase/models/permissions.clj [hunk 1]
```diff
@@ -487,12 +487,24 @@
...
-           (classloader/require ' metabase-enterprise.advanced-permissions.models.permissions)
-           (resolve ' metabase-enterprise.advanced-permissions.models.permissions/data-model-write-perms-path))]
-    (if (and f premium-features/enable-advanced-permissions?)
+           (classloader/require 'metabase-enterprise.advanced-permissions.models.permissions)
+           (resolve 'metabase-enterprise.advanced-permissions.models.permissions/data-model-write-perms-path))]
+    (if (and f (premium-features/enable-advanced-permissions?))
       (f db-id schema table-id)
       "/")))
 
+(s/defn db-details-write-perms-path :- Path
```
Comment (noahmoss) [inline src/metabase/models/permissions.clj:496]: This function is basically the same as the one above for data model perms, but I'm going to experiment using a macro to abstract this stuff away in a future PR. (see [this thread](https://metaboat.slack.com/archives/CKZEMT1MJ/p1649176488140789) for context)

File: src/metabase/models/permissions.clj [hunk 2]
```diff
@@ -487,12 +487,24 @@
   permissions path for the table. Otherwise, defaults to the root path ('/'), thus restricting writes to admins."
   [db-id schema table-id]
   (let [f (u/ignore-exceptions
-           (classloader/require ' metabase-enterprise.advanced-permissions.models.permissions)
-           (resolve ' metabase-enterprise.advanced-permissions.models.permissions/data-model-write-perms-path))]
-    (if (and f premium-features/enable-advanced-permissions?)
+           (classloader/require 'metabase-enterprise.advanced-permissions.models.permissions)
+           (resolve 'metabase-enterprise.advanced-permissions.models.permissions/data-model-write-perms-path))]
+    (if (and f (premium-features/enable-advanced-permissions?))
```
Comment (noahmoss) [inline src/metabase/models/permissions.clj:492]: 🤦 this needed to be a function call. added a couple tests that would have caught this.

**Comment by alxnddr** [comment]
@noahmoss what is the name of the prop for checking if a user has access to the database details page? Similar to `can_access_data_model`

**Comment by noahmoss** [comment]
@alxnddr oops, I forgot that part, sorry. just updated the PR to add it as `can_access_db_details`

**Review by howonlee — APPROVED** [review]
I eagerly await the macro idea, I got burned by this too somewhere

File: enterprise/backend/src/metabase_enterprise/advanced_permissions/models/permissions.clj [hunk 3]
```diff
@@ -123,7 +123,9 @@
 ;;; +----------------------------------------------------------------------------------------------------------------+
 
 (defn data-model-write-perms-path
-  "Returns the permissions path required to edit the data model for a table specified by `path-components`."
+  "Returns the permissions path required to edit the data model for a table specified by `path-components`.
+  This is a simple wrapper around `perms/feature-perms-path`, but it lives in an EE namespace to ensure that data model
+  permissions only work when EE code can be loaded."
```
[the remark this answers is not in the document]
Comment (noahmoss) [reply enterprise/backend/src/metabase_enterprise/advanced_permissions/models/permissions.clj:128]: If EE code isn't loaded, it shouldn't explode -- it should fallback to requiring full admin permissions.

This is essentially tested by the existing tests in the OSS namespace for `/api/field`, `/api/table`, etc.

**Pull request merged**